src/node/node_state.h

```diff
@@ -841,9 +841,10 @@ namespace ccf
 
       auto timer_msg = std::make_unique<threading::Tmsg<NodeStateMsg>>(
         [](std::unique_ptr<threading::Tmsg<NodeStateMsg>> msg) {
+          std::lock_guard<pal::Mutex> guard(msg->data.self.lock);
           if (msg->data.self.sm.check(NodeStartupState::pending))
           {
-            msg->data.self.initiate_join();
+            msg->data.self.initiate_join_unsafe();
```

Comment:
TSAN caught one failure here:

Without the lock, we first branch on `check(pending)`, and then inside `initiate_join` the first thing we do is `require(pending)`. If a join succeeds between the check and the require, the require fails and the node dies. This resolves that by locking the access. I'd be tempted to flatten the `check` into `initiate_join()`, but this fix is smaller.